R/thumbnail.R

```diff
@@ -0,0 +1,327 @@
+#' Get content item thumbnail
+#'
+#' Download the thumbnail for a content item on Connect to a file on your
+#' computer.
+#'
+#' @param content A content item.
+#' @param path Optional. A path to a file used to write the thumbnail image. If
+#' no path is provided, a temporary file with the correct file extension is
+#' created. If the content item does not have a thumbnail, returns `NA`.
+#' 
+#' @returns The path to the image file.
+#' 
+#' @examples
+#' \dontrun{
+#' client <- connect()
+#' item <- content_item(client, "8f37d6e0-3395-4a2c-aa6a-d7f2fe1babd0")
+#' thumbnail <- get_thumbnail(item)
+#' }
+#'
+#' @family thumbnail functions
+#' @family content functions
+#' @export
+get_thumbnail <- function(content, path = NULL) {
+  validate_R6_class(content, "Content")
+  guid <- content$content$guid
+  con <- content$connect
+
+  # Connect 2024.09.0 introduced public endpoints for content thumbnails. We
+  # prefer those, falling back to the unversioned endpoints if unavailable.
+  res <- con$GET(
+    url = con$server_url("content", guid, "__thumbnail__"),
+    parser = NULL
+  )
+  if (httr::status_code(res) == 404) {
+    res <- con$GET(
+      unversioned_url("applications", guid, "image"),
+      parser = NULL
+    )
+  }
+  con$raise_error(res)
+
+  if (httr::status_code(res) == 204) {
+    return(NA_character_)
+  }
+
+  # Guess file extension
+  if (is.null(path)) {
+    ct <- httr::headers(res)$`content-type`
+    if (grepl("^image/", ct)) {
+      # Just strip off 'image/'
+      ext <- names(mime::mimemap[mime::mimemap == ct])[1]
+      path <- tempfile(pattern = "content_image_", fileext = paste0(".", ext))
+    } else {
+      # Try png
+      warning(glue::glue("Could not infer file extension from content type: {ct}. Using '.png'"))
```

Comment:
Why would we ever be here? Does Connect serve non-images, or not emit Content-Type? And why would assuming png be better than assuming nothing? 

How about this for the logic here?

```
if (is.null(path)) {
  path <- tempfile(pattern = "content_image_")
}

ct <- httr::headers(res)$`content-type`
if (ct %in% mime::mimemap) {
  ext <- names(mime::mimemap[mime::mimemap == ct])[1]
  if (!endsWith(path, ext) {
    path <- paste(path, ext, sep = ".")
  }
}
```

That way, (a) we never assume a wrong type, and (b) we make sure that `path` always ends with the right file extension, even if I provide a mismatched one, so that `httr` will do the right thing with it on upload (since that was the reason you identified for trying to identify the content type in the first place).

Reply:
Yeah, I was already considering using the determined extension to override the user-provided extension — I like that.

I don't know why the original code uses `png` in the case of content type being indeterminate. Perhaps a better behavior would be to leave the file with no extension and emit a warning; do you like that better?

Reply:
One wrinkle with this approach verbatim is that you can easily get `tempfile.jpg.jpeg` 😂